modules/component/src/core/cd-aware/strategy.ts

```diff
@@ -0,0 +1,330 @@
+// @Notice this part of the code is in the coalescing PR
+// import { generateFrames } from '../projections/generateFrames';
+// import { coalesce } from '../operators/coalesce';
+import { defer, from, MonoTypeOperatorFunction, Observable } from 'rxjs';
+import {
+  ChangeDetectorRef,
+  NgZone,
+  ɵdetectChanges as detectChanges,
+  ɵmarkDirty as markDirty,
+} from '@angular/core';
+import { hasZone, isIvy } from '../utils';
+import { mapTo } from 'rxjs/operators';
+
+function getSaveDurationSelector(ngZone: NgZone): () => Observable<number> {
+  return () => defer(() => from(Promise.resolve()).pipe(mapTo(1)));
```

Comment:
You can save on promise allocations here... but it's mildly tricky, because we have to wait for Zones to patch the Promise.... so change this like so:

```ts
/** A shared promise instance to cause a delay of one microtask */
let resolvedPromise: Promise<void>|null = null;

function getResolvedPromise() {
  resolvedPromise = resolvedPromise || Promise.resolve();
  return resolvedPromise;
}
function getSaveDurationSelector(ngZone: NgZone): () => Observable<number> {
  return () => defer(getResolvedPromise).pipe(mapTo(1));
```

Alternatively, this will do the same as above, just more tersely and with more promise allocations:

```ts
return () => defer(() => Promise.resolve(1));
```

Reply:
AFAICT zone patched EVERYTHING before Angular bootstraps. 

I can get access to the unpatched version of a promise over:
```typescript
(window as any).__zone_symbol__Promise
```
for the `animationFrame` it is:

```typescript
 (window as any).__zone_symbol__requestAnimationFrame;
 (window as any).__zone_symbol__cancelAnimationFrame;
```

So I would check if we are in zone full env and return the patched version or if zoneless the original like this for Promises:
```typescript
/** A shared promise instance to cause a delay of one microtask */
let resolvedPromise: Promise<void>|null = null;

function getResolvedPromise() {
  resolvedPromise = resolvedPromise ||  hasZone(ngZone) ? (window as any).__zone_symbol__Promise.resolve() : Promise.resolve();
  return resolvedPromise;
}

function getSaveDurationSelector(ngZone: NgZone): () => Observable<number> {
    defer(getResolvedPromise).pipe(mapTo(1));
}
```

And for animation frame this:
```typescript
function getAnimationFrame() {
 return hasZone(ngZone)
      // custom operator similar to animationFrames, just let me pass the reference to the function to generate frames from (set/cancel interval, timeout, animationFrame)
      ? generateFrames(
          (window as any).__zone_symbol__requestAnimationFrame,
          (window as any).__zone_symbol__cancelAnimationFrame
        )
      : animationFrames();
}

function getSaveDurationSelector(ngZone: NgZone): () => Observable<number> {
    defer(getAnimationFrame);
}
```


Here the last question:

Promise vs AnimationFrame I had no differences in the tests but would go for AnimationFrame. At first, I thought I could miss values, but then I equalized angular is also using the AnimationFrame to coalesce.

Any technical debt here?

Reply:
Implemented.